### ISSUE TITLE
[$1000] Web -  Onyx: out-of-order updates

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#28737]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
This bug comes from discussion in  https://github.com/Expensify/App/issues/26592
TL;DR: Onyx processes the `successData` and API response in the wrong order on iOS and Android.
Dive deeper:
https://github.com/Expensify/App/blob/3352cc8ecaa80390fe37c09ee20f8bbf7bec09d7/src/libs/actions/Policy.js#L408
When adding a member to workspace here, we send the following data:
```
"optimisticData": {
    "onyxMethod": "set",
    "key": "report_3008335418540078",
    "value": {
        "type": "chat",
        "chatType": "policyExpenseChat",
        "ownerAccountID": 899964195,
        "participantAccountIDs": [
            15664021,
            899964195
        ],
        "policyID": "41A8E180FAEB43DF",
        "reportID": "3008335418540078",
        "reportName": "Chat Report",
        "pendingFields": {
            "createChat": "add"
        },
        "isOptimisticReport": true
        ...
    }
},
"successData": {
    "onyxMethod": "merge",
    "key": "report_3008335418540078",
    "value": {
        "pendingFields": {
            "createChat": null
        },
        "errorFields": {
            "createChat": null
        },
        "isOptimisticReport": false
    }
}
```
And the API call returns the following onyx `merge_collection` action:
```
{
    "jsonCode": 200,
    "requestID": "806fecea0afd3512-WAW",
    "onyxData": [
         {
            "onyxMethod": "mergecollection",
            "key": "report_",
            "value": {
                ...
                "report_3008335418540078": {
                    "reportID": "3008335418540078",
                    "reportName": "",
                    "type": "chat",
                    "chatType": "policyExpenseChat",
                    "ownerEmail": "[pasha@test.co](mailto:pasha@test.co)",
                    "ownerAccountID": 15671904,
                    "policyID": "41A8E180FAEB43DF",
                    "participants": [
                        "[pasha@test.co](mailto:pasha@test.co)",
                        "[paultsimura+3@gmail.com](mailto:paultsimura+3@gmail.com)"
                    ],
                    "participantAccountIDs": [
                        15664021,
                        15671904
                    ],
                    "lastActorAccountID": 15671904,
                },   
            }
        },
    ]
}
```
It should replace the temporary `accountID` of the `report.ownerAccountID` with the real one, which comes from the server.
On the web, the onyx actions are executed in the following order:
1. persist optimisticData;
2. persist successData;
3. persist API response;
The issue is that on iOS, the onyx actions are executed in the following order:
1. persist optimisticData;
2. **persist API response;**
3. **persist successData;**
The `successData` holds merge operation of only several certain fields, which shouldn't affect the report much.
**However,** here's the tricky part:
This piece of code, which creates a promise for future merge operation, is executed before Onyx persists the API response:
https://github.com/Expensify/react-native-onyx/blob/616c4c5286b7a7fadac18d477cd75edd60865b12/lib/Onyx.js#L1136-L1147
Meaning the following execution chain:
1. persist optimisticData;
2. create `persistSuccessData` promise;
3. persist API response;
4. execute `persistSuccessData` promise;
This means, at the moment `persistSuccessData` is created, the existingValue is still not updated, so even though it should update only a couple of fields, the `modifiedData` here is saved in a state of pre-API-response.
Then the report is updated by the "persist API response" operation, but gets overwritten by the `persistSuccessData` operation with the outdated data.
If we debug step-by-step, you'll notice that the chat name is displayed correctly between the steps 3 and 4 from above.
As an outcome of this complete operation, the state of Onyx is the following:
- personalDetailsList contains the new permanent ownerAccountID;
- report has the temporary ownerAccountID, which is already removed from Onyx.

## Expected Result:
Onyx updates are executed in correct order without data being overwritten

## Actual Result:
 Onyx updates face a race condition and some data is overwritten

## Workaround:
Unknown 

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [ ] Android / Chrome
- [x] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.76.0
**Reproducible in staging?:** n/a
**Reproducible in production?:** n/a
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

n/a

**Expensify/Expensify Issue URL:**
**Issue reported by:** @@paultsimura
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1696287156456449

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~013373cff08a2575e9</li>
        <li>Upwork Job ID: 1709846441891180544</li>
        <li>Last Price Increase: 2023-10-05</li>
    </ul>
</details>

**Comment by ospfranco** [comment]
I believe I found the cause.

When the `update` function is called, all the operations [are fired in an async manner](https://github.com/Expensify/react-native-onyx/blob/main/lib/Onyx.js#L1433). However, this causes a race condition where certain operations will be written into the DB before others. It's a built-in race condition. 

Just for documentation purposes: I suggested we need to make the update operations sequentially await. @marcaaron @tgolen are however worried about the app taking a big performance hit. I will try another solution and report back if I manage to get the app to do reliable updates.

**Comment by pradeepmdk** [comment]
> I believe I found the cause.
> 
> When the `update` function is called, all the operations [are fired in an async manner](https://github.com/Expensify/react-native-onyx/blob/main/lib/Onyx.js#L1433). However, this causes a race condition where certain operations will be written into the DB before others. It's a built-in race condition.
> 
> Just for documentation purposes: I suggested we need to make the update operations sequentially await. @marcaaron @tgolen are however worried about the app taking a big performance hit. I will try another solution and report back if I manage to get the app to do reliable updates.

@ospfranco for note: that is what I already told in my proposal. https://github.com/Expensify/App/issues/26592#issuecomment-1736981022
cc: @Julesssss

**Comment by pradeepmdk** [comment]
![Screenshot 2023-10-05 at 10 46 05 AM](https://github.com/Expensify/App/assets/16595846/d1c019a3-504c-4fcd-ab37-020506ba1e46)

https://github.com/Expensify/App/assets/16595846/736946e5-183f-4b16-969a-c0ef118ad228
![Screenshot 2023-10-05 at 11 26 13 AM](https://github.com/Expensify/App/assets/16595846/eb38fdc4-0471-404b-86fe-e75ecda43816)

@marcaaron I saw recently it changed to onyx to memory here 
https://github.com/Expensify/App/pull/25614

### Issue here:
1) we are flushed out to Onyx once all the api is completed in the queue (networkRequestQueue).` I think this will create a performance issue right?` ex: here in the queue `2035` is pending after coming online it gets a response object in the `queuedOnyxUpdates` to `46k` object.

2) https://github.com/Expensify/App/pull/25614 Due to this change incase the middle of the response I refresh the page means it will never update to onyx.

### solution:
[App/src/libs/Network/SequentialQueue.js](https://github.com/Expensify/App/blob/41ddd7428340fd8d347e8160980809e76a93899b/src/libs/Network/SequentialQueue.js#L126)
 we need to flush out each api success or failure.

cc: @marcaaron @tgolen @ospfranco @mountiny

**Comment by Julesssss** [comment]
@CortneyOfstad bumping the price as this one is pretty critical

**Comment by CortneyOfstad** [comment]
@jjcoffee any thoughts on the comments/proposals above?

**Comment by jjcoffee** [comment]
Unsure if this is still reproducible based on [this](https://github.com/Expensify/App/issues/26592#issuecomment-1747963504), but also looks like @ospfranco has a [PR](https://github.com/Expensify/react-native-onyx/pull/384) in the works?

**Comment by pradeepmdk** [comment]
@jjcoffee https://github.com/Expensify/App/issues/26592#issuecomment-1747963504, is fixed . but still have some issues can you check my https://github.com/Expensify/App/issues/28737#issuecomment-1748146420

**Comment by CortneyOfstad** [comment]
Bump @jjcoffee ^^^ Thanks!

**Comment by jjcoffee** [comment]
Sorry @pradeepmdk could you flesh out your proposal or try to explain it a bit better? I'm finding it hard to follow what new issue you've found.

**Comment by ospfranco** [comment]
From what I understand API requests responses are placed in a networkRequestQueue, which was before persisted to disk but now it is only kept in memory. If one refreshes the tab, then this queue will be lost (also due to its size it will probably take some time to process). Which could potentially lead to an inconsistent state. IMO this is another issue, not even sure if it is an issue if it doesn't cause incosistent data in the app, and should be in a separate ticket.

[my PR](https://github.com/Expensify/App/issues/28737) is merged and should have already fixed the out-of-order database writes. Let me know if there is anything else.

**Comment by Julesssss** [comment]
Hi @ospfranco, the PR link you shared just links back to this issue.

**Comment by ospfranco** [comment]
Oops, sorry, this is the PR

https://github.com/Expensify/react-native-onyx/pull/384

**Comment by pradeepmdk** [comment]
>From what I understand API requests responses are placed in a networkRequestQueue, which was before persisted to disk but now it is only kept in memory. If one refreshes the tab, then this queue will be lost (also due to its size it will probably take some time to process). Which could potentially lead to an inconsistent state. IMO this is another issue, not even sure if it is an issue if it doesn't cause incosistent data in the app, and should be in a separate ticket.

@ospfranco  Thank you

@Julesssss  @jjcoffee what do you think about this

**Comment by Julesssss** [comment]
> IMO this is another issue, not even sure if it is an issue if it doesn't cause incosistent data in the app, **and should be in a separate ticket.**

Agree

**Comment by ospfranco** [comment]
Unfortunately my PR has been reverted. It was causing other issues: https://github.com/Expensify/react-native-onyx/pull/397 

Please hold on from closing/testing this ticket until further notice. We are discussing with the team how to best tackle the issue.

**Comment by paultsimura** [comment]
I think this finding by @tienifr might also be helpful while testing: https://github.com/Expensify/App/issues/28941#issuecomment-1757511553

**Comment by MonilBhavsar** [comment]
> Please hold on from closing/testing this ticket until further notice. We are discussing with the team how to best tackle the issue.

Could you please link the discussion thread. Thanks!

**Comment by CortneyOfstad** [comment]
@ospfranco Bump on Monil's comment above?

**Comment by ospfranco** [comment]
https://expensify.slack.com/archives/C035J5C9FAP/p1697031295068549?thread_ts=1696934662.824599&cid=C035J5C9FAP

**Comment by jjcoffee** [comment]
@CortneyOfstad I believe this one is due payment (PR hit [production](https://github.com/Expensify/App/pull/29169#issuecomment-1805970076) 10th Nov).

**Comment by paultsimura** [comment]
@Julesssss could you please also clarify what was the plan on this payment?
https://github.com/Expensify/App/issues/26592#issuecomment-1742664307

**Comment by Julesssss** [comment]
Hi @paultsimura, could you please clarify your question? Matt is in charge of payment for that issue

**Comment by paultsimura** [comment]
Sure @Julesssss 
It was a loong time ago, but in the linked comment, you wrote:
> @paultsimura create a new bug for the out-of-order updates that you discovered, we can continue that investigation and pay out for your help separately from this issue

That was because I discovered that out-of-order behavior of Onyx and we decided to make it a separate issue.

So I just wanted to clarify how that should happen

**Comment by paultsimura** [comment]
Thank you, that's solid

**Comment by jjcoffee** [comment]
@CortneyOfstad Friendly bump for payment! The PR hit [production](https://github.com/Expensify/App/pull/29169#issuecomment-1805970076) 10th Nov so this is overdue. :bow:

**Comment by paultsimura** [comment]
@jjcoffee as per the comment above, I believe we are just waiting for your and @ospfranco 's confirmation first:

> Lets just seek confirmation from these three before paying out

**Comment by jjcoffee** [comment]
Ah okay, sounds good to me, @ospfranco are you okay with [this](https://github.com/Expensify/App/issues/28737#issuecomment-1818689981)?

**Comment by Julesssss** [comment]
Sure. @CortneyOfstad could you pay out @paultsimura and @jjcoffee. And @ospfranco if we can locate their details.

**Comment by paultsimura** [comment]
@CortneyOfstad I received a reporter contract. Could you please recheck the amount from this comment: https://github.com/Expensify/App/issues/28737#issuecomment-1818689981

**Comment by CortneyOfstad** [comment]
@paultsimura It shows in the original comment that you are the reporter of the issue. We do not pay the other amount unless you are the contributor or reviewer, which are @jjcoffee and @ospfranco. 

Bug reports are $50 flat rate.

**Comment by Julesssss** [comment]
Hey @CortneyOfstad, this was an exception where @paultsimura highlighted a big problem with Onyx that we were then able to fix. At the time, the bug bounty was still in place, but I view this as a wider architectural problem that would have caused us big problems in the future. For these reasons, we decided to pay a larger bounty.

**Comment by Julesssss** [comment]
@CortneyOfstad yeah 👍

**Comment by CortneyOfstad** [comment]
Sounds good @Julesssss!

@paultsimura I've adjusted the offer in Upwork, so please let me know once you accept!

@ospfranco I still need you to link me your Upwork profile so I can send you an offer. Thanks!

**Comment by paultsimura** [comment]
Thank you @CortneyOfstad, accepted the offer

**Comment by CortneyOfstad** [comment]
Actually, can't find them in Slack 🤔

**Comment by shubham1206agra** [comment]
@Julesssss Can you tell me if this issue is fixed or not? Cause I saw the linked PR and I am not able to understand how changes to `withOnyx` solved the race condition in `Onyx.update`.

**Comment by shubham1206agra** [comment]
@Julesssss Bump on the question ^

**Comment by Julesssss** [comment]
There are explanations in the [original](https://github.com/Expensify/react-native-onyx/pull/355) and follow up [onyx repo PR](https://github.com/Expensify/react-native-onyx/pull/385)

**Comment by Julesssss** [comment]
> Actually, can't find them in Slack 🤔

I found them on Twitter, but I got blocked for typing a message that sounded like a scam 😆 

lets see if they get my second message

**Comment by ospfranco** [comment]
ping :)

**Comment by ospfranco** [comment]
Ah, so actually my PR was reverted, I don't know which steps the team took to fix the issue if any at all. As it had entailed doing a large refactoring of the Onyx internals. I cannot in good conscience accept any payment since I'm no longer helping Margelo with the project. You can forgo any payment you are planning for me, it's fine :)

**Comment by Julesssss** [comment]
@ospfranco thanks, I see. It's hard to tell but it looks like we would have reverted, then resubmitted with a small fix which would still be eligible for payment. 

But I'll leave it up to you

**Comment by CortneyOfstad** [comment]
@ospfranco bump on the comment above ^^^ Thanks!

**Comment by Julesssss** [comment]
Sounds like he's happy not to. We can always reopen if that changes though!
